### PR TITLE
BABE SRML tests

### PR DESCRIPTION
**Pull request by DemiMarie-parity**
As requested by @rphmeier.  Based on #3305.

### CODE REVIEW

File: srml/babe/src/mock.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline srml/babe/src/mock.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: srml/babe/src/lib.rs [hunk 2]
```diff
@@ -17,13 +17,13 @@
 //! Consensus extension module for BABE consensus.
 
 #![cfg_attr(not(feature = "std"), no_std)]
-#![forbid(unused_must_use, unsafe_code, unused_variables, dead_code)]
-
+#![forbid(unused_must_use, unsafe_code, unused_variables)]
+#![deny(dead_code)]
```
Comment (bkchr) [inline srml/babe/src/lib.rs]:
```suggestion
```

File: core/consensus/babe/primitives/src/digest.rs [hunk 3]
```diff
@@ -17,7 +17,7 @@
 //! Private implementation details of BABE digests.
 
 #[cfg(feature = "std")]
-use super::AuthoritySignature;
+use super::sr25519::AuthoritySignature;
```
Comment (bkchr) [inline core/consensus/babe/primitives/src/digest.rs]: That is not correct. You should use the app signature.

File: core/consensus/babe/primitives/src/lib.rs [hunk 4]
```diff
@@ -30,22 +30,27 @@ use substrate_client::decl_runtime_apis;
...
 	use app_crypto::{app_crypto, key_types::BABE, sr25519};
 	app_crypto!(sr25519, BABE);
 }
 
-/// A Babe authority keypair. Necessarily equivalent to the schnorrkel public key used in
-/// the main Babe module. If that ever changes, then this must, too.
-#[cfg(feature = "std")]
-pub type AuthorityPair = app::Pair;
+/// Sr25519 crypto for BABE
+pub mod sr25519 {
```
Comment (bkchr) [inline core/consensus/babe/primitives/src/lib.rs]: What is the reason for this change? Babe only works with `sr25519`?

Comment (rphmeier) [reply core/consensus/babe/primitives/src/lib.rs]: I suppose to make it easier to mock. BABE could work with other kinds of crypto, but it's hard to reason about the necessary generalizations we'd have to make.

Comment (DemiMarie-parity) [reply core/consensus/babe/primitives/src/lib.rs]: @rphmeier Yup.  This way, I can use the existing `UintAuthorityId` infrastructure.

File: srml/babe/src/tests.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	with_externalities(&mut new_test_ext(vec![0, 1, 2, 3]), || {
+		assert_eq!(Babe::authorities().len(), 4)
+	})
+}
+
+#[test]
+fn check_module() {
+	with_externalities(&mut new_test_ext(vec![0, 1, 2, 3]), || {
+		assert!(!Babe::should_end_session(0), "Genesis does not change sessions");
+		assert!(!Babe::should_end_session(200000),
```
Comment (kianenigma) [inline srml/babe/src/tests.rs]: should be indented.

File: srml/babe/src/tests.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+type System = system::Module<Test>;
+type Session = session::Module<Test>;
+type EpochDuration = <Test as super::Trait>::EpochDuration;
+
+#[test]
+fn check_epoch_change() {
+	with_externalities(&mut new_test_ext(vec![0, 1, 2, 3]), || {
+		System::initialize(&1, &Default::default(), &Default::default(), &Default::default());
+		assert!(!Babe::should_end_session(0), "Genesis does not change sessions");
+		assert!(!Babe::should_end_session(200000),
```
Comment (kianenigma) [inline srml/babe/src/tests.rs]: likewise: if one is wrapped to next line then best to it for all.

File: srml/babe/src/tests.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+			"BABE does not include the block number in epoch calculations");
+	})
+}
+
+type System = system::Module<Test>;
+type Session = session::Module<Test>;
+type EpochDuration = <Test as super::Trait>::EpochDuration;
+
+#[test]
+fn check_epoch_change() {
```
Comment (kianenigma) [inline srml/babe/src/tests.rs]: I cannot understand what this test is doing but can recomend that by organizing it a bit better (simple as empty line between logical blocks) and some commenting and preferably breaking it into smaller ones with the simple format of: 

```
// given 
some_setup()

// when
some_event()

// then
some_assertion()
```

it would be much easier to read and argue. RN it really looks like a big blob of code to me. 
```

Comment (DemiMarie-parity) [reply srml/babe/src/tests.rs]: Hopefully fixed.  Please let me know it is not.

File: srml/babe/Cargo.toml [hunk 8]
```diff
@@ -11,6 +11,7 @@ serde = { version = "1.0.93", optional = true }
 inherents = { package = "substrate-inherents", path = "../../core/inherents", default-features = false }
 rstd = { package = "sr-std", path = "../../core/sr-std", default-features = false }
 sr-primitives = { path = "../../core/sr-primitives", default-features = false }
+sr-version = { path = "../../core/sr-version", default-features = false }
```
Comment (andresilva) [inline srml/babe/Cargo.toml]: Should this go under `dev-dependencies`? Only seems to be used for the mock.

File: srml/babe/src/lib.rs [hunk 9]
```diff
@@ -42,9 +39,15 @@ use inherents::{RuntimeString, InherentIdentifier, InherentData, ProvideInherent
...
 use inherents::{InherentDataProviders, ProvideInherentData};
 use babe_primitives::{BABE_ENGINE_ID, ConsensusLog, BabeAuthorityWeight, Epoch, RawBabePreDigest};
-pub use babe_primitives::{AuthorityId, VRF_OUTPUT_LENGTH, PUBLIC_KEY_LENGTH};
+pub use babe_primitives::{sr25519::AuthorityId, VRF_OUTPUT_LENGTH, PUBLIC_KEY_LENGTH};
 use system::ensure_root;
 
+#[cfg(all(feature = "std", test))]
+mod tests;
+
+#[cfg(all(test, feature = "std"))]
```
Comment (andresilva) [inline srml/babe/src/lib.rs]:
```suggestion
#[cfg(all(feature = "std", test))]
```

Nitpick, but the different ordering here and above was triggering my OCD.

File: srml/babe/src/tests.rs [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Consensus extension module tests for BABE consensus.
+#![allow(unused_imports)]
+#![cfg(test)]
+use super::*;
+use runtime_io::with_externalities;
+use mock::{new_test_ext, Babe, Test};
+use sr_primitives::{traits::Header, Digest, testing::UintAuthorityId};
+use babe_primitives::{CompatibleDigestItem, sr25519::AuthorityId};
```
Comment (andresilva) [inline srml/babe/src/tests.rs]: Unused?

File: core/consensus/babe/primitives/src/lib.rs (hunk 4 above)
Comment (andresilva) [reply core/consensus/babe/primitives/src/lib.rs]: I still don't understand why this is needed. We are doing a `pub use *` of everything in this module just below.

File: srml/babe/src/tests.rs (hunk 10 above)
Comment (DemiMarie-parity) [reply srml/babe/src/tests.rs]: yeah

**Comment by DemiMarie-parity** [comment]
@rphmeier @andresilva can you have another look?

File: srml/babe/src/tests.rs [hunk 11]
```diff
@@ -95,7 +95,7 @@ fn check_epoch_change() {
 
 		// Check that we got the expected digest.
 		let Digest { ref logs } = header.digest;
-		assert_eq!(logs.len(), 1, "should only have 1 digest here");
+		assert_eq!(logs.len(), 2, "should have exactly 2 digests here");
```
Comment (rphmeier) [inline srml/babe/src/tests.rs]: what are those 2 digests? Is the change from 1 -> 2 right or was it just made to make the tests pass?

Comment (DemiMarie-parity) [reply srml/babe/src/tests.rs]: Just to make tests pass, sadly.  I will check.

Comment (rphmeier) [reply srml/babe/src/tests.rs]: If it really should be 2 digests, that's fine. But it's not good to change things in tests just to get the green checkmark when it should be red.

Comment (DemiMarie-parity) [reply srml/babe/src/tests.rs]: @rphmeier Right now I am not sure what it should be, tbh.

**Comment by gavofyork** [comment]
> As requested by @rphmeier. 

Then I guess he's best placed to review :)

**Comment by rphmeier** [comment]
Yup, getting it in shape now.

**Comment by DemiMarie-parity** [comment]
Superseded by #3652.
